### PR TITLE
chore: send package version as agent header once per new client

### PR DESCRIPTION
**Pull request by anitarua**
addresses #85 

kind of hacky approach to trying to get the version number from the SDK's pubspec.yaml actually doesn't work when I try importing the package locally into another Dart project

### CODE REVIEW

File: lib/src/internal/control_client.dart [hunk 1]
```diff
@@ -28,15 +33,33 @@ class ControlClient implements AbstractControlClient {
...
+  Future<Map<String, String>> makeHeaders({String? cacheName}) async {
+    var headers = <String, String>{};
+    if (cacheName != null) {
+      headers.addEntries({'cache': cacheName}.entries);
+    }
+    if (firstRequest) {
+      firstRequest = false;
+      try {
+        String? packageVersion = await findPackageVersion();
+        headers.addEntries({'agent': packageVersion ?? 'unknown'}.entries);
```
Comment (cprice404) [inline lib/src/internal/control_client.dart:45]: if i'm reading this right, the header would only include the version number; is that accurate? we want to make sure it includes the word "dart". even if the version number is unknown.

but lmk if i'm reading the code wrong!

Comment (anitarua) [reply lib/src/internal/control_client.dart:45]: oh yeah, you're right, I can add "dart" in there

File: lib/src/internal/data_client.dart [hunk 2]
```diff
@@ -52,15 +55,35 @@ class DataClient implements AbstractDataClient {
...
+            metadata: {'authorization': credentialProvider.apiKey},
+            timeout: _configuration.transportStrategy.grpcConfig.deadline));
+  }
+
+  Future<Map<String, String>> makeHeaders({String? cacheName}) async {
+    var headers = <String, String>{};
+    if (cacheName != null) {
+      headers.addEntries({'cache': cacheName}.entries);
+    }
+    if (firstRequest) {
```
Comment (cprice404) [inline lib/src/internal/data_client.dart:75]: might be worth pulling this whole clause into your utils namespace? not a blocker.

File: lib/src/internal/utils/utils.dart [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  File? pubspecFile = await findPubspecFile();
+  if (pubspecFile != null) {
+    Map pubspec = loadYaml(pubspecFile.readAsStringSync());
+    return pubspec['version'];
+  }
+  return null;
+}
+
+// Construct relative path to the pubspec.yaml file
+Future<File?> findPubspecFile() async {
```
Comment (cprice404) [inline lib/src/internal/utils/utils.dart:18]: when a project is pulling the momento lib as a binary artifact dependency from pub.dev, do you know how that works? do you expect this yaml file to be in place there?

File: lib/src/internal/utils/utils.dart [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import 'dart:isolate';
+
+import 'package:yaml/yaml.dart';
+
+// Extract the package version number from
+// the pubspec.yaml file if it's found
+Future<String?> findPackageVersion() async {
+  File? pubspecFile = await findPubspecFile();
+  if (pubspecFile != null) {
+    Map pubspec = loadYaml(pubspecFile.readAsStringSync());
```
Comment (cprice404) [inline lib/src/internal/utils/utils.dart:11]: what you are doing is probably fine, but just curious if you have seen [this](https://pub.dev/documentation/pubspec_parse/latest/pubspec_parse/Pubspec/Pubspec.parse.html)

**Review by cprice404 — APPROVED** [review]
This seems fine to try out! i left a few nits but they are not blockers. we'll want to make sure we test it from a toy consumer app once there is a version of this published to pub.dev.

number

File: lib/src/internal/utils/utils.dart (hunk 3 above)
Comment (anitarua) [reply lib/src/internal/utils/utils.dart:18]: I actually just tried this and the yaml file is not there and the paths I get are not useful for finding it, so I might just set the header to a hardcoded string for now

```
Current directory path: /Users/anita/cli
Current script path: /Users/anita/cli/.dart_tool/pub/bin/cli/cli.dart-3.2.3.snapshot
INFO: 2024-01-12 10:28:14.893626: Setting agent header to dart:unkown
```

File: lib/src/internal/utils/utils.dart (hunk 4 above)
Comment (anitarua) [reply lib/src/internal/utils/utils.dart:11]: I think that one converts yaml string to a pubspec file

File: lib/src/internal/utils/utils.dart (hunk 3 above)
Comment (cprice404) [reply lib/src/internal/utils/utils.dart:18]: you could always put the version number in a standalone source file that had nothing in it except for a constant, and then update the release please logic to make it bump that version number in addition to the one in the pubspec.yaml.

Comment (cprice404) [reply lib/src/internal/utils/utils.dart:18]: https://github.com/googleapis/release-please/blob/main/docs/customizing.md#updating-arbitrary-files

Comment (cprice404) [reply lib/src/internal/utils/utils.dart:18]: https://github.com/googleapis/release-please/blob/main/src/updaters/generic.ts

**Review by cprice404 — APPROVED** [review]
❤️

**Pull request merged**